packages/lesswrong/components/posts/usePostReadProgress.ts

```diff
@@ -32,45 +33,65 @@ interface UsePostReadProgressProps {
-    const totalHeight = useFirstViewportHeight ? postBodyElement.getBoundingClientRect().height : window.scrollY + postBodyBottomPos;
-  
-    const scrollPercent = 1 - ((postBodyBottomPos + delayStartOffset) / totalHeight);
-    const adjustedScrollPercent = clamp ? clampPct(scrollPercent) : scrollPercent;
-    return adjustedScrollPercent * 100;
+  /**
+   * Calculates what percentage of the post has been scrolled through
+   * Uses the bottom of the viewport as the reference point
+   * @returns A value between 0-100 representing percentage of post that's been scrolled, unless clamp=false
+   */
+  function getScrollPct(
+    postBodyElement: HTMLElement,
+    clamp = true
+  ): number {
+    const containerTop = getOffsetChainTop(postBodyElement); 
+    const containerHeight = postBodyElement.offsetHeight;
+    const bottomOfViewport = window.scrollY + window.innerHeight;
+
+    const distanceFromPostTopToViewportBottom = bottomOfViewport - containerTop;
+    const scrollPercent = (distanceFromPostTopToViewportBottom / containerHeight) * 100;
+
+    if (clamp) {
+      return clampPct(scrollPercent);
+    }
+    return scrollPercent;
   }
   
+  /**
+   * What percent of the post body's height is revealed within the window viewport.
+   * We want to show "scroll window" that's proportionally large, relative to the height of the entire progress bar container.
+   * @returns The height the scrollbar indicator should be in pixels
+   */
   const getScrollWindowHeight = (postBodyElement: HTMLElement, readingProgressBarContainerElement: HTMLElement) => {
-    const postBodyHeight = postBodyElement.getBoundingClientRect().height;
-    const windowHeight = window.innerHeight;
+    const postRect = postBodyElement.getBoundingClientRect();
+    const viewportHeight = window.innerHeight;
+    const tocHeight = readingProgressBarContainerElement.getBoundingClientRect().height;
+    
+    // If post is smaller than viewport, the window should be 100% of the TOC height
+    if (postRect.height <= viewportHeight) {
+      return { displayedWindowScrollHeight: tocHeight };
+    }
     
-    /**
-     * What percent of the post body's height is revealed within the window viewport.
-     * We want to show "scroll window" that's proportionally large, relative to the height of the entire progress bar container.
-     */
-    const windowBodyPercent = clampPct(windowHeight / postBodyHeight) * 100;
-    const containerHeight = readingProgressBarContainerElement.getBoundingClientRect().height;
-    const absoluteScrollWindowHeight = (windowBodyPercent / 100) * containerHeight;
-    const currentScrollPct = getScrollPct(postBodyElement, delayStartOffset);
+    // Calculate how much of the post is actually visible in the viewport
+    const visibleTop = Math.max(0, Math.min(viewportHeight, postRect.top));
+    const visibleBottom = Math.max(0, Math.min(viewportHeight, postRect.bottom));
+    const visibleHeight = Math.max(0, visibleBottom - visibleTop);
```

Comment:
Are any of the second values here ever going to be negative? I'm confused by what the Math.max is doing if not

Reply:
possRect.top and postRect.bottom can be both be negative, indicating they have been scrolled past (extend past top of the viewport.

Reply:
Oh, huh, ok.